parse/src/lib.rs

```diff
@@ -0,0 +1,143 @@
+
+    let mut string_it = string.chars();
+    let mut maybe_ch = string_it.next();
+    let mut num: f64 = 0.0;
+    let mut neg_found = false;
+
+    match maybe_ch {
+        Some('-') => {
+            neg_found = true;
+            maybe_ch = string_it.next();
+        }
+        Some('+') => {
+            maybe_ch = string_it.next();
+        }
+        None => {
+            return Err(pfe_empty());
+        }
+        _ => {}
+    }
+
+    // Process digits before the decimal point.
+    while let Some(ch) = maybe_ch {
+        match ch {
+            '0'...'9' => {
+                num *= 10.0;
+                num += ((ch as u8) - ('0' as u8)) as f64;
+            }
+            '.' | 'e' | 'E' => {
+                break;
+            }
+            _ => {
+                return Err(pfe_invalid());
+            }
+        }
+        maybe_ch = string_it.next();
+    }
+
+    // Process digits after the decimal point
+    if maybe_ch == Some('.') {
+        let mut mul = 1.0;
+        maybe_ch = string_it.next();    // get first character after '.'
+        while let Some(ch) = maybe_ch {
+            match ch {
+                '0'...'9' => {
+                    mul *= 0.1;
+                    num += (((ch as u8) - ('0' as u8)) as f64) * mul;
+                }
+                'e' | 'E' => {
+                    break;
+                }
+                _ => {
+                    return Err(pfe_invalid());
+                }
+            }
+            maybe_ch = string_it.next();
+        }
+    }
+
+    // Process exponent
+    if maybe_ch == Some('e') || maybe_ch == Some('E') {
+        let mut neg_exp_found = false;
+        maybe_ch = string_it.next();    // Get character after 'e'
+        match maybe_ch {
+            Some('-') => {
+                neg_exp_found = true;
+                maybe_ch = string_it.next();
+            }
+            Some('+') => {
+                maybe_ch = string_it.next();
+            }
+            _ => {}
+        }
+        let mut exp = 0;
+        while let Some(ch) = maybe_ch {
+            match ch {
+                '0'...'9' => {
+                    exp *= 10;
+                    exp += ((ch as u8) - ('0' as u8)) as i32;
+                }
+                _ => {
+                    return Err(pfe_invalid());
+                }
+            }
+            maybe_ch = string_it.next();
+        }
+        if neg_exp_found {
+            exp = -exp;
+        }
+        num *= 10_f64.powi(exp);
+    }
+
+    if maybe_ch.is_none() {
+        if neg_found {
+            num = -num;
+        }
+        Ok(num)
+    } else {
+        Err(pfe_invalid())
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use ::parse_f64;
+
+    fn test_f64(string: &str) {
+        let tst_num = parse_f64(string);
+        let ref_num = string.parse::<f64>();
+
+        println!("{:?} {:?}", tst_num, ref_num);
+        assert_eq!(format!("{:?}", tst_num), format!("{:?}", ref_num));
```

Comment:
Because with floating point, the internal representation could be different, even though it prints the same. So I'd have to compare with an epsilon. All I really care about is whether its the same as far as significant digits are concerned.

Also parse::ParseFloatError is not the same struct as std::num::ParseFloatError (which is republished core::num::ParseFloatError, so I'd have to write special tests. Comparing the strings seemed easier.